python/mxnet/gluon/estimator/estimator.py

```diff
@@ -236,32 +231,128 @@ def fit(self, train_data,
                 # update metrics
                 for metric in self.metrics:
                     metric.update(label, pred)
-                    self.train_stats['batch_' + metric.name] = metric.get()[1]
+                    # get metric name and current value and update train stats
+                    self.train_history.set_batch_stats(*metric.get())
                 for loss, loss_metric, in zip(losses, self.loss_metrics):
                     loss_metric.update(0, [l for l in loss])
-                    self.train_stats['batch_' + loss_metric.name] = loss_metric.get()[1]
+                    self.train_history.set_batch_stats(*loss_metric.get())
 
-                try:
-                    self.train_stats['step'] = "{}/{}".format(batch_size * (i + 1), len(train_data._dataset))
-                except AttributeError:
-                    self.train_stats['step'] = i
+                self.train_history.batch_idx = i
+                # record trained samples v.s. total samples if using Gluon DataLoader
+                if isinstance(train_data, gluon.data.DataLoader):
+                    self.train_history.samples = "{}/{}".format(batch_size * (i + 1), len(train_data._dataset))
 
-                for trainer in self.trainers:
-                    trainer.step(batch_size)
+                self.trainer.step(batch_size)
 
                 # batch end
                 for handler in event_handlers:
                     handler.batch_end()
 
             for metric in self.metrics + self.loss_metrics:
-                self.train_stats['train_' + metric.name].append(metric.get()[1])
+                self.train_history.set_train_stats(*metric.get())
             # epoch end
             for handler in event_handlers:
                 handler.epoch_end()
 
-            if self.stop_training:
+            if self.train_history.stop_training:
                 break
 
         # train end
         for handler in event_handlers:
             handler.train_end()
+
+
+class TrainHistory(object):
```

Comment:
I think you should make this extensible without needing to derive this class for additional stats?